### ISSUE TITLE
__GNUC_PREREQ is not portable

### ISSUE THREAD

**Issue by citrus-it** [cyrusimap/cyrus-imapd#3462]
I'm just trying to upgrade cyrus imapd to 3.4.0 on an OmniOS system and am getting build errors due to `__GNUC_PREREQ` not being defined. From what I can tell, this is a macro provided by `features.h` on systems which use glibc, and therefore not portable.

This is the patch I'm using temporarily to get around this:

```diff
-#  if defined(__GNUC__) && __GNUC_PREREQ(4, 3)
+#  if defined(__GNUC__) && ((__GNU__ == 4 && __GNUC_MINOR >= 3) || __GNU__ > 4)
```

**Comment by elliefm** [comment]
Thanks, that looks helpful

**Comment by elliefm** [comment]
@brong Our references to __GNUC_PREREQ were introduced by you, in 63cae46e04e and 10785fbaf40.  Does this patch look like a reasonable expression of the same intent?

Though it looks like we're going for a gcc- (and maybe glibc-) specific optimisation here, so making the macro portable enough that "things that are not gcc but pretend to be" end up in the same path might be asking for more trouble...

@citrus-it Does something like this help, instead?

```diff
-#  if defined(__GNUC__) && __GNUC_PREREQ(4, 3)
+#  if defined(__GNUC__) && defined(__GNUC_PREREQ) && __GNUC_PREREQ(4, 3)
```

**Comment by citrus-it** [comment]
> @citrus-it Does something like this help, instead?

That would likely make it compile but would just remove the optimisation for anybody not using glibc, which is probably not the intention. For example, https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html documents ` __builtin_bswap32()`, so that one is safe to use as long as it's gcc.

I've seen other projects add something like this early in the headers so that __GNUC_PREREQ is available everywhere:

```C
#ifndef __GNUC_PREREQ
# if defined __GNUC__ && defined __GNUC_MINOR__
#  define __GNUC_PREREQ(maj, min) \
	((__GNUC__ << 16) + __GNUC_MINOR__ >= ((maj) << 16) + (min))
# else
#  define __GNUC_PREREQ(maj, min) 0
# endif
#endif /* __GNUC_PREREQ */
```

**Comment by elliefm** [comment]
> That would likely make it compile 

Are you able to try it and confirm please?

> but would just remove the optimisation for anybody not using glibc, which is probably not the intention.

Actually that _was_ my intention (but not necessarily the original intention).  If the thing claiming to be __GNUC__ doesn't provide __GNUC_PREREQ, then I don't feel like it can be fully trusted to support whatever tricky thing we were trying to do, and so it's better to skip the optimisation and use the standard fallback.

In this specific case of `__builtin_bswap32()`, it seems like it's probably fine, but I feel like providing our own __GNUC_PREREQ is going to be a headache to debug when the next not-quite-compatible quirk comes up.

**Comment by citrus-it** [comment]
> > That would likely make it compile
> 
> Are you able to try it and confirm please?

Yes, it compiles with that change.

> > but would just remove the optimisation for anybody not using glibc, which is probably not the intention.
> Actually that _was_ my intention (but not necessarily the original intention).

Ok, sorry for misunderstanding that.

> If the thing claiming to be `__GNUC__` doesn't provide `__GNUC_PREREQ`, then I don't feel like it can be fully trusted to support whatever tricky thing we were trying to do

That's the problem - `gcc` is the thing claiming to be `__GNUC__`, but `__GNUC_PREREQ()` comes from the header files from the system C library. By relying on it, you're effectively checking for gcc in conjunction with "Linux with glibc" (and some *BSDs that have it as a compatibility macro). Even Linux systems which use musl or uclibc are excluded as far as I can tell.

If you actually just want to test for gcc >= 4.3, please consider using something like:

```C
#if defined(__GNUC__) && ((__GNU__ == 4 && __GNUC_MINOR >= 3) || __GNU__ > 4)
```

**Comment by elliefm** [comment]
This is now fixed on master and 3.4 branches, and will be included in the 3.4.1 release.  Thanks :)
